src/cryptonote_core/service_node_list.cpp

```diff
@@ -404,6 +522,33 @@ namespace service_nodes
     return true;
   }
 
+  static int count_bits(unsigned char byte)
+  {
+    return byte == 0 ? 0 : 1 + count_bits(byte & (byte-1));
+  }
+
+  uint64_t service_node_list::get_swarm_id_for_pubkey(const crypto::public_key& pubkey) const
+  {
+    std::vector<uint64_t> swarm_ids = get_swarm_ids();
+    std::vector<std::pair<size_t, uint64_t> > swarm_ids_with_distance(swarm_ids.size());
+    std::pair<size_t, uint64_t> best = std::make_pair(1024, 0);
+    for (size_t i = 0; i < swarm_ids.size(); i++)
+    {
+      char data[32];
+      memcpy(data + 0 * sizeof(swarm_ids[i]), &swarm_ids[i], sizeof(swarm_ids[i]));
+      memcpy(data + 1 * sizeof(swarm_ids[i]), &swarm_ids[i], sizeof(swarm_ids[i]));
+      memcpy(data + 2 * sizeof(swarm_ids[i]), &swarm_ids[i], sizeof(swarm_ids[i]));
+      memcpy(data + 3 * sizeof(swarm_ids[i]), &swarm_ids[i], sizeof(swarm_ids[i]));
+
+      size_t distance = 0;
+      for (size_t i = 0; i < sizeof(data); i++)
+        distance += count_bits(((unsigned char)data[i]) ^ (unsigned char)pubkey.data[i]);
+
+      best = std::min(best, std::make_pair(distance, swarm_ids[i]));
```

Comment:
Looks complicated. Is there a reference for this algorithm? Curious to read more

Reply:
you're looking at a reference, read the code. it counts the number of bits that are similar

the concept is called hamming distance https://en.wikipedia.org/wiki/Hamming_distance